### PR TITLE
fix(@desktop/wallet): creating a common grouped model for accounts asets

### LINKED ISSUE #12601: explore the wallet account assets service and model
This task is being used to get a single groupedModel for account assets that can be used in Wallet views , Send, Profile and TokenManagement places. 

This task will only do this for the WalletView
the remaining places will be handled in different PRs-

- [ ] #13016
- [ ] 2. Profile Layout
- [ ] 3. Token Management
- [ ] 4. Remove old tokens_service under accounts service and only keep new one. 
- [ ] 5. clean up old tokens model which is not being used anywhere else 
- [ ] #13084

### PR DESCRIPTION
**Pull request by Khushboo-dev-cpp**
fixes #12601

### What does the PR do

Creates a grouped account assets model in the service :
```
 {
            tokensKey: "DAI",
            balances: [
                { account: "0x7F47C2e18a4BBf5487E6fb082eC2D9Ab0E6d7240", chainId: 5, balance: "0"},
                { account: "0x7F47C2e98a4BBf5487E6fb082eC2D9Ab0E6d8881", chainId: 5, balance: "0"}
            ]
        },
        {
            tokensKey: "ETH",
            balances: [
                {account: "0x7F47C2e18a4BBf5487E6fb082eC2D9Ab0E6d7240", chainId: 420, balance: "1013151281976507736"},
                {account: "0x7F47C2e18a4BBf5487E6fb082eC2D9Ab0E6d7240", chainId: 421613, balance: "473057568699284613"},
                {account: "0x7F47C2e18a4BBf5487E6fb082eC2D9Ab0E6d7240", chainId: 5, balance: "307400931315122839"},
                {account: "0x7F47C2e98a4BBf5487E6fb082eC2D9Ab0E6d8881", chainId: 420, balance: "122082928968121891"},
                {account: "0x7F47C2e98a4BBf5487E6fb082eC2D9Ab0E6d8881", chainId: 421613, balance: "0"},
                {account: "0x7F47C2e98a4BBf5487E6fb082eC2D9Ab0E6d8881", chainId: 5, balance: "559133758939097000"}
            ]
        }
```

this model is then passed on by nim via the `grouped_account_assets_model.nim` to the qml.
This implements the new approach that all other data need for tokens will be fetched from the tokensBySymbolModel and Community Model using unique keys and a LeftJoinModel.

Also sorting , filtering will be UIs responsibility including filtering based on account selection and networks on the main Wallet View.

This PR only uses this new approach in the main AssetsView and AssetDetailsView.
It will be used in Profile, SendModal and TokenManagement under this task https://github.com/status-im/status-desktop/issues/13016.

Currently we have 3 calls from client to get market values,
1.  from the old getWalletTokens api which will be remove in task mentioned above
2. from the new tokens service which still uses getWalletTokens to be updated in this task https://github.com/status-im/status-desktop/issues/13084
3. From all tokens service which will remain. 

### Affected areas

AssetsView.qml, AssetDetailsView.qml
### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [ ] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [ ] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

https://github.com/status-im/status-desktop/assets/60327365/7f56ed93-0b6a-4e22-88b7-31bd7df6ec33

https://github.com/status-im/status-desktop/assets/60327365/f4bd3690-f903-42fb-b162-cd03eaec1b20

### CODE REVIEW

File: storybook/pages/AssetsViewPage.qml
Comment (caybro) [inline storybook/pages/AssetsViewPage.qml:1]: Ouch, I just created it here as well: https://github.com/status-im/status-desktop/pull/12837

File: ui/imports/shared/controls/TokenDelegate.qml [hunk 1]
```diff
@@ -18,14 +18,14 @@ StatusListItem {
 
     property string currentCurrencySymbol
     property string textColor: {
-        if (!modelData) {
+        if (!modelData || !modelData.marketDetails) {
```
Comment (caybro) [inline ui/imports/shared/controls/TokenDelegate.qml:21]: This is going to conflict as well :/

File: src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim [hunk 2]
```diff
@@ -129,26 +130,53 @@ QtObject:
...
+  proc tokensDetailsAboutToUpdate*(self: FlatTokensModel) =
+    if self.delegate.getFlatTokensList().len > 0:
+      let index = self.createIndex(0, 0, nil)
+      let lastindex = self.createIndex(self.delegate.getFlatTokensList().len-1, 0, nil)
       defer: index.delete
-      self.dataChanged(index, index, @[ModelRole.MarketDetails.int, ModelRole.MarketDetailsLoading.int])
+      defer: lastindex.delete
+      self.dataChanged(index, lastindex, @[ModelRole.Description.int, ModelRole.WebsiteUrl.int, ModelRole.DetailsLoading.int])
 
   proc tokensDetailsUpdated*(self: FlatTokensModel) =
```
Comment (caybro) [inline src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim:167]: I can't see any difference between this method and the above (`tokensDetailsAboutToUpdate`); what's the deal here?

File: src/app/modules/main/wallet_section/all_tokens/market_details_item.nim [hunk 3]
```diff
@@ -23,64 +23,72 @@ QtObject:
...
     result.symbol = symbol
-    result.currencyFormat = delegate.getCurrentCurrencyFormat()
 
   proc updateCurrencyFormat*(self: MarketDetailsItem) =
     self.currencyFormat = self.delegate.getCurrentCurrencyFormat()
 
   proc marketCapChanged*(self: MarketDetailsItem) {.signal.}
   proc marketCap*(self: MarketDetailsItem): QVariant {.slot.} =
-    return newQVariant(self.delegate.getMarketValuesBySymbol(self.symbol).marketCap)
+    if self.symbol.isEmptyOrWhitespace or self.delegate.getTokensMarketValuesLoading(): return newQVariant("")
```
Comment (caybro) [inline src/app/modules/main/wallet_section/all_tokens/market_details_item.nim:32]: Why returning an empty string here? Wouldn't just `newQVariant()` do?

Comment (caybro) [reply src/app/modules/main/wallet_section/all_tokens/market_details_item.nim:32]: Same below

File: ui/imports/shared/views/AssetsView.qml [hunk 4]
```diff
@@ -251,27 +293,29 @@ ColumnLayout {
...
+                return LocaleUtils.currencyAmountToLocaleString(root.currencyStore.getCurrentCurrencyAmount(totalCurrencyBalance))
             }
             errorMode: !!networkConnectionStore ? networkConnectionStore.noBlockchainConnectionAndNoCache && !networkConnectionStore.noMarketConnectionAndNoCache : false
             errorIcon.tooltip.text: !!networkConnectionStore ? networkConnectionStore.noBlockchainConnectionAndNoCacheText : ""
             onClicked: (itemId, mouse) => {
                            if (mouse.button === Qt.LeftButton) {
-                               RootStore.getHistoricalDataForToken(modelData.symbol, RootStore.currencyStore.currentCurrency)
+                               RootStore.getHistoricalDataForToken(modelData.symbol, root.currencyStore.currentCurrency)
                                d.selectedAssetIndex = delegateIndex
-                               let selectedView = !!modelData.communityId ? communityAssetsLV : regularAssetsLV
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:273]: Why did you revert this?

File: ui/imports/shared/views/AssetsView.qml [hunk 5]
```diff
@@ -282,8 +326,7 @@ ColumnLayout {
             Component.onCompleted: {
                 // on Model reset if the detail view is shown, update the data in background.
                 if(root.assetDetailsLaunched && delegateIndex === d.selectedAssetIndex) {
-                    let selectedView = !!modelData.communityId ? communityAssetsLV : regularAssetsLV
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:285]: Same, why reverted?

File: ui/imports/shared/controls/TokenDelegate.qml [hunk 6]
```diff
@@ -96,7 +96,7 @@ StatusListItem {
             StatusTextWithLoadingState   {
                 id: currencyBalance
                 anchors.right: parent.right
-                text: modelData ? LocaleUtils.currencyAmountToLocaleString(modelData.enabledNetworkCurrencyBalance) : ""
```
Comment (caybro) [inline ui/imports/shared/controls/TokenDelegate.qml:99]: Where has this value gone, what's the text now when it's not loading?

File: ui/imports/shared/views/AssetsView.qml [hunk 7]
```diff
@@ -62,42 +69,78 @@ ColumnLayout {
             settingsKey: "WalletAssets"
         }
 
-        readonly property bool hasCommunityAssets: communityAssetsLV.count
-
         function hideAllCommunityTokens(communityId) {
-            const tokenSymbols = ModelUtils.getAll(communityAssetsLV.model, "symbol", "communityId", communityId)
+            const tokenSymbols = ModelUtils.getAll(regularAssetsLV.model, "symbol", "communityId", communityId)
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:73]: Why the change here?

File: ui/imports/shared/views/AssetsView.qml [hunk 8]
```diff
@@ -62,42 +69,78 @@ ColumnLayout {
...
+                        root.addressFilters
+                        return root.networkFilters.split(":").includes(chainId+"") &&
+                                (!!root.addressFilters ? root.addressFilters.toUpperCase() === account.toUpperCase() : true)
+                    }
+                }
+            }
+        }
 
-        sourceModel: root.assets
+        function getTotalBalance(balances, decimals) {
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:94]: Ugh really do we want to sum the numbers by using JS on the client side?

File: ui/imports/shared/views/AssetsView.qml [hunk 9]
```diff
@@ -62,42 +69,78 @@ ColumnLayout {
...
+                 expression: !!model.communityId
+                 expectedRoles: ["communityId"]
             }
         ]
         filters: [
             ExpressionFilter {
                 expression: {
-                    d.controller.settingsDirty
-                    return d.tokenIsVisible(model.symbol, model.currentCurrencyBalance) && (customFilter.isCommunity ? !!model.communityId : !model.communityId)
+                    if (!!model.communityId) {
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:132]: No, we want to filter also non-community tokens

File: ui/imports/shared/views/AssetsView.qml [hunk 10]
```diff
@@ -217,16 +269,6 @@ ColumnLayout {
                     onClicked: Global.openPopup(communityInfoPopupCmp)
                 }
             }
-
-            StatusListView {
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:221]: I don't understand, why is this removed?

**Review by caybro — COMMENTED** [review]
Some comments inline, will test this locally later; impressive amount of work but I have some doubts, esp. regarding the Settings/Token management features

File: ui/imports/shared/views/AssetsView.qml [hunk 11]
```diff
@@ -62,42 +69,78 @@ ColumnLayout {
...
+    property SortFilterProxyModel customSFPM: SortFilterProxyModel {
+        sourceModel: d.assetsWithFilteredBalances
         proxyRoles: [
-            ExpressionRole {
-                name: "currentBalance"
-                expression: model.enabledNetworkBalance.amount
+                FastExpressionRole {
+                id: filter
+                 name: "currentBalance"
+                 expression: d.getTotalBalance(model.balances, model.decimals, root.addressFilters, root.networkFilters)
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:109]: What if the `balances` or the filters change? This being a JS function will not trigger reevaluation

File: ui/imports/shared/views/AssetsView.qml (hunk 9 above)
Comment (caybro) [reply ui/imports/shared/views/AssetsView.qml:132]: The original expression:
```js
(customFilter.isCommunity ? !!model.communityId : !model.communityId)
```
is different

File: src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim (hunk 2 above)
Comment (Khushboo-dev-cpp) [reply src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim:167]: ok so the tokensDetailsAboutToUpdate is emitted when refetch the details from the service via an async call and tokensDetailsUpdated when the details are fetched. So we update the model accordingly. Here we rely on data from service and dont hold our own copies of the data and hence this is needed. This will also help updating UI state to loading/loaded as needed.

File: ui/imports/shared/views/AssetsView.qml (hunk 4 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:273]: Hey @caybro Iv basically updated the logic here from having 2 listviews to have a single listview with sections instead. Its no reverted simply updated . Same for all the comments with regards to these changes

File: ui/imports/shared/views/AssetsView.qml (hunk 5 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:285]: see answer above

File: ui/imports/shared/controls/TokenDelegate.qml (hunk 6 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/controls/TokenDelegate.qml:99]: its assigned from the outside now in AssetsView.qml

File: ui/imports/shared/views/AssetsView.qml (hunk 7 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:73]: https://github.com/status-im/status-desktop/pull/12871#discussion_r1434200352

File: ui/imports/shared/views/AssetsView.qml (hunk 10 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:221]: https://github.com/status-im/status-desktop/pull/12871#discussion_r1434200352

File: ui/imports/shared/views/AssetsView.qml (hunk 8 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:94]: I know, its not ideal but we tried to use the SumAggregator here but it wont work, this works fairly quickly, until we have a better solution for this.

File: ui/imports/shared/views/AssetsView.qml (hunk 11 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:109]: hey so, me adding the filters in this function call root.addressFilters and root.networkFilters re-triggers this expression role.

File: ui/imports/shared/views/AssetsView.qml (hunk 9 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:132]: yes I am aware, you needed that for the 2 list views but now that we have a single one, 
`(customFilter.isCommunity ? !!model.communityId : !model.communityId)` is not needed.
The reason for having condition for the community is that we check for currency balance, but currency balance for community tokens will always be 0 as they have no real price on any market, hence they should not be filtered out by this filter.

File: ui/imports/shared/views/AssetsView.qml (hunk 8 above)
Comment (caybro) [reply ui/imports/shared/views/AssetsView.qml:94]: I see, probably not a big deal as the amount of assets is kinda limited/finite :)

File: ui/imports/shared/views/AssetsView.qml (hunk 9 above)
Comment (caybro) [reply ui/imports/shared/views/AssetsView.qml:132]: No, that's not the point; this function `d.tokenIsVisible()` also checks for those explicitly hidden by the user (in Manage tokens). Now you're letting those slip

File: src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim (hunk 2 above)
Comment (caybro) [reply src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim:167]: Yeah but still... the implementation is the same? That's what looked fishy :)

File: ui/imports/shared/views/AssetsView.qml (hunk 8 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:94]: yes!

File: src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim (hunk 2 above)
Comment (Khushboo-dev-cpp) [reply src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim:167]: aah yes you are right, we are doing the same thing both places, I could add a function to avoid code repetition but it is intentional

**Review by alaibe — APPROVED** [review]
Approving assuming @caybro comments will be fixed

File: ui/imports/shared/views/AssetsView.qml (hunk 9 above)
Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml]: gotcha! Ive kept the filter for community tokens too now :)

Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml]: @caybro Ive fixed all your comments now, please check when you can :)

File: ui/imports/shared/views/AssetsView.qml [hunk 12]
```diff
@@ -175,33 +215,42 @@ ColumnLayout {
         verticalPadding: 0
         contentWidth: availableWidth
 
+        StatusListView {
+            id: regularAssetsLV
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:219]: Maybe also worth renaming this one, now that it serves both categories

File: ui/imports/shared/views/AssetsView.qml [hunk 13]
```diff
@@ -175,33 +215,42 @@ ColumnLayout {
...
-                objectName: "assetViewStatusListView"
-                model: CustomSFPM {}
-                delegate: delegateLoader
-            }
-
             StatusDialogDivider {
                 Layout.fillWidth: true
                 Layout.topMargin: Style.current.padding
                 Layout.bottomMargin: Style.current.halfPadding
-                visible: d.hasCommunityAssets
```
Comment (caybro) [inline ui/imports/shared/views/AssetsView.qml:196]: This divider should still be visible iff there are some community assets

Comment (Khushboo-dev-cpp) [reply ui/imports/shared/views/AssetsView.qml:196]: as this is a part of the section divider already we dont need to have this condition anymore
